crates/ruff_python_formatter/README.md

```diff
@@ -0,0 +1,306 @@
+`ruff check`.
+
+In future releases, the Ruff formatter will likely support configuration of quote style (single vs.
+double) and indentation (width, and spaces vs. tabs).
+
+### Suppressing formatting
+
+Ruff supports Black's `# fmt: off`, `# fmt: on`, and `# fmt: skip` pragmas, with a few caveats.
+
+See Ruff's [suppression comment proposal](https://github.com/astral-sh/ruff/discussions/6338) for
+details.
+
+## Black compatibility
+
+The formatter is designed to be a drop-in replacement for [Black](https://github.com/psf/black).
+
+Specifically, the formatter is intended to emit near-identical output when run over Black-formatted
+code. When migrating an existing project from Black to Ruff, you should expect to see a few
+differences on the margins, but the vast majority of your code should be formatted identically.
+Note, however, that the formatter does not yet implement or support Black's preview style.
+
+When run over _non_-Black-formatted code, the formatter makes some different decisions than Black,
+and so more deviations should be expected.
+
+### Intentional deviations
+
+This section enumerates the known, intentional deviations between the Ruff formatter and Black's
+stable style. (Unintentional deviations are tracked in the [issue tracker](https://github.com/astral-sh/ruff/issues?q=is%3Aopen+is%3Aissue+label%3Aformatter).)
+
+#### Trailing end-of-line comments
+
+Black's priority is to fit an entire statement on a line, even if it contains end-of-line comments.
+In such cases, Black collapses the statement, and moves the comment to the end of the collapsed
+statement:
+
+```python
+# Input
+while (
+    cond1  # almost always true
+    and cond2  # almost never true
+):
+    print("Do something")
+
+# Black
+while cond1 and cond2:  # almost always true  # almost never true
+    print("Do something")
+```
+
+Ruff, like [Prettier](https://prettier.io/), expands any statement that contains trailing
+end-of-line comments. For example, Ruff would avoid collapsing the `while` test in the snippet
+above. This ensures that the comments remain close to their original position and retain their
+original intent, at the cost of retaining additional vertical space.
+
+This deviation only impacts unformatted code, in that Ruff's output should not deviate for code that
+has already been formatted by Black.
+
+### Pragma comments are ignored when computing line width
+
+Pragma comments (`# type: ignore`, `# noqa`, etc.) are ignored when computing the width of a line.
+This prevents Ruff from moving pragma comments around, thereby modifying their meaning and behavior:
+
+See Ruff's [pargma comment handling proposal](https://github.com/astral-sh/ruff/discussions/6670)
+for details.
+
+This is similar to [Pyink](https://github.com/google/pyink) but a deviation from Black. Black uses
+this behavior for `# type: ignore` comments ([#997](https://github.com/psf/black/issues/997)), while
```

Comment:
Black's behavior for `type: ignore` is different: It avoids splitting any line that has a type ignore comment, limiting formatting mainly to normalizing spaces.